### PR TITLE
"include_all_instances", "exclude_not_covering_patterns" and "print_only_repair_results" options added.

### PR DESCRIPTION
**Pull request by khaes-kth**
Hi,

Three options are added.

1&2- "include_all_instances_for_each_tool" & "exclude_repair_patterns_not_covering_the_whole_diff" that can be set by "-parameter include_all_instances_for_each_tool:true". The first one causes the output to include all instances of each tool (not only one for each tool). The second one (if set true) excludes the instancePatterns that do not cover the whole diff. see: [link](https://github.com/khaes-kth/coming/blob/master/src/main/java/fr/inria/coming/repairability/RepairabilityAnalyzer.java#L69)

3- "print_only_repair_results" which can be set in the same way. If set true, repairability mode returns the results of RepairabilityAnalyzer not PatternInstanceAnalyzer. This one is a temporary fix for #172 
 see: [link](https://github.com/khaes-kth/coming/blob/master/src/main/java/fr/inria/coming/repairability/JSONRepairabilityOutput.java#L32)

### CODE REVIEW

**Comment by monperrus** [comment]
Excellent. Could you document those options in the appropriate MD file in the `docs` folder?

**Comment by martinezmatias** [comment]
Hi @khaes-kth 
Thanks, that's a nice feature. Some comments:

- We will need some test cases to test each case.

- About the implementation of the method `coversTheWholeDiff`: it's a fine first implementation but I have some doubt (which can be removed having testing scenarios). The doubts are two: 1) the comparison of Operations (not sure which fields are compared and what happens when ops duplicates) 2) the use of AllOperations from the diff (rather than RootOps).
I was thinking about another way to do this check. Note that there are severals. A new implementation could leverage on the mapping information that each ChangePatter Instance has i.e., that links Patterns Specification's  elements with those from the real diff. This information is [HERE](https://github.com/SpoonLabs/coming/blob/HEAD/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/ChangePatternInstance.java#L26).
Thus, one strategy could be to check whether operations from the Diff are mapped or not. If there is not not mapped it means that it's not part of the pattern instance.
WDYT?
Regards
Matias

**Comment by khaes-kth** [comment]
Hi @martinezmatias,

1- Tests: I'm adding the tests. However, before it we should reach a solution for the other point you have mentioned.

2- About coversTheWholeDiff: 
Why `allOperations `instead of RootOps? As you can see [HERE](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/DetectorChangePatternInstanceEngine.java#L378), the tool finds the matching actions among allOperations and not rootOps. I thought this means that in a pattern_instance that covers the whole diff, allOperations should be mapped.

Why this way of comparison (`diffOperation.toString().equals(instanceOperation.toString()`)? Because diffOperation iterates over all operations in the diff and instanceOperation iterates over all mapped operations. Therefore, if all instances are matched and `toString()` function generates a unique string for each operation and this string is always the same for an operation, this is the correct way to compare them. I'm not sure but I supposed that `toString()` does this.

Regarding your recommendation for implementation: "one strategy could be to check whether operations from the Diff are mapped or not". Yes, that's exactly what I have tried to do. I think `[actions](https://github.com/SpoonLabs/coming/blob/HEAD/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/ChangePatternInstance.java#L27)` also includes the list of mapped operations (the same as `actionOperation.valueSet()`). If this is correct, my implementation works the same as your recommended version.

**Comment by martinezmatias** [comment]
Hi @khaes-kth 

> I thought this means that in a pattern_instance that covers the whole diff, allOperations should be mapped.

Yes, that's right. Covering all ops should be enough.

> Why allOperations instead of RootOps? As you can see HERE, the tool finds the matching actions among allOperations and not rootOps. 

The reason of using AllOps instead of Roots is as follows:
Imagine that you have a diff:
```
+ if (a> b){
+   i = 0;}
```

The root will have one Operation (Insert If) , the AllOp will have 2 (Insert If, Insert Assign)
In the matching we consider all ops because we can be interested in finding , for instance, all ops that affects Assignment.

> Regarding your recommendation for implementation: "one strategy could be to check whether operations from the Diff are mapped or not". Yes, that's exactly what I have tried to do. I think actions also includes the list of mapped operations (the same as actionOperation.valueSet()). If this is correct, my implementation works the same as your recommended version.

Great!


> Why this way of comparison (diffOperation.toString().equals(instanceOperation.toString())?

I am not sure what the toString shows. Consequently, I would use other ways to compare (method Equals, or even a tailored comparator which considers the fields we want to compare)

**Comment by khaes-kth** [comment]
Hi again,

> Covering all ops should be enough.
> 
> >
Yes, that's enough. But due to your explanation, I think it is not necessary. It seems that covering RootOps is enough as well. Consider your example and suppose that we have a repair pattern that inserts an if_block. In this case, the repair pattern contains only one action/operation, while diff.allOperations contains two operations. So one of the elements in allOperations would not be mapped, although the repair tool might be able to reproduce the diff. Note that the INSERT_ASSIGNMENT action does not have to be included in the pattern. [repairTool.filter()](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/repairability/repairtools/Arja.java#L64) can filter the diffs with a mapped RootOp that cannot be reproduced by the tool.
WDYT?

> I would use other ways to compare

After reading the equals() function, I think it does what we need. Also, I tested it on [uCrop ](https://github.com/Yalantis/uCrop) project and `diffOperation.equals(instanceOperation)` worked the same as `diffOperation.toString().equals(instanceOperation.toString())`. I will use equals() instead of toString().

**Comment by monperrus** [comment]
@khaes-kth  will push the tests today

File: src/main/java/fr/inria/coming/repairability/RepairabilityAnalyzer.java [hunk 1]
```diff
@@ -81,4 +92,19 @@ public AnalysisResult analyze(IRevision input, RevisionResult previousResults) {
 
         return finalResult;
     }
+
+    private boolean coversTheWholeDiff(ChangePatternInstance instancePattern, Diff diff) {
+        for(Operation diffOperation : diff.getAllOperations()){
```
Comment (monperrus) [inline src/main/java/fr/inria/coming/repairability/RepairabilityAnalyzer.java]: per the discussion on Github and today in person, we change this to getRootOperations

**Comment by martinezmatias** [comment]
Hi @khaes-kth 
Is this feature ready to merge?
Thanks!
Matias

**Comment by khaes-kth** [comment]
Hi @martinezmatias 


> Is this feature ready to merge?

Yes, I believe so.

**Comment by martinezmatias** [comment]
Okey, I merge, thanks!

**Comment by khaes-kth** [comment]
Thank you.

**Comment by monperrus** [comment]
Thanks for the effective progress!

**Pull request merged**